### PR TITLE
Similarity component siamese network

### PR DESCRIPTION
**Pull request by clemenszuck**
**Description**
Adds a new similarity component including two new API-routes:

* http://localhost:8003/similarity to calculate a similarity matrix for a list of embeddings

* http://localhost:8003/train to re-train the network on a list of labeled embedding pairs

**Testing**
* Start up Athene using `docker-compose up -d`
* Test if the API routes work as expected (example JSONs and expected results in [similarity/README.md](https://github.com/ls1intum/Athene/blob/similarity-component-siamese-network/similarity/README.md))

### CODE REVIEW

File: similarity/src/SimilarityResource.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        # Create a JSON representation of the resource
+        resp.body = json.dumps(doc, ensure_ascii=False, default=self.__default)
+
+        # The following line can be omitted because 200 is the default
+        # status returned by the framework, but it is included here to
+        # illustrate how this may be overridden as needed.
+        resp.status = HTTP_200
+        self.__logger.info("Completed Similarity Request.")
+        self.__logger.debug("-" * 80)
```
Comment (linusmichel) [inline similarity/src/SimilarityResource.py:65]: Newline at EOF missing

File: similarity/src/entities.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    label: int
+
+    def __init__(self, embedding1: Embedding, embedding2: Embedding, label: int):
+        self.embedding1 = embedding1
+        self.embedding2 = embedding2
+        self.label = label
+
+    @classmethod
+    def from_dict(cls, dict: dict) -> 'EmbeddingsPair':
+        return cls(dict['embedding1'], dict['embedding2'], dict['label'])
```
Comment (linusmichel) [inline similarity/src/entities.py:46]: Newline at EOF missing

File: similarity/src/errors.py [hunk 3]
```diff
@@ -0,0 +1,9 @@
+from falcon import HTTPBadRequest
+
+emptyBody = HTTPBadRequest("Empty Requerst Body", "Cannot process request with empty body.")
+requireTwoBlocks = HTTPBadRequest("Need two or more blocks", "Must provide at least two text blocks.")
+requireCourseId = HTTPBadRequest("CourseId not provided", "Must provide a course identifier to upload material")
+requireFileData = HTTPBadRequest("File data is not provided", "Must provide file data to upload material")
+requireFileName = HTTPBadRequest("File name is not provided", "Must provide file name to upload material")
+requireTwoEmbeddings = HTTPBadRequest("Need two or more embeddings", "Most provide at least two embeddings.")
+requireEmbeddingPairs = HTTPBadRequest("No embedding pairs provided.", "Must provide embedding pairs to train network.")
```
Comment (linusmichel) [inline similarity/src/errors.py:9]: Newline at EOF missing

File: similarity/src/resources/siamese-model-default.json [hunk 4]
```diff
@@ -0,0 +1 @@
...": 0.0010000000474974513}}, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}}]}, "inbound_nodes": [[["input_1", 0, 0, {}]], [["input_2", 0, 0, {}]]]}, {"name": "dot_1", "class_name": "Dot", "config": {"name": "dot_1", "trainable": true, "dtype": "float32", "axes": 1, "normalize": true}, "inbound_nodes": [[["sequential_1", 1, 0, {}], ["sequential_1", 2, 0, {}]]]}, {"name": "dense_2", "class_name": "Dense", "config": {"name": "dense_2", "trainable": true, "dtype": "float32", "units": 1, "activation": "sigmoid", "use_bias": true, "kernel_initializer": {"class_name": "VarianceScaling", "config": {"scale": 1.0, "mode": "fan_avg", "distribution": "uniform", "seed": null}}, "bias_initializer": {"class_name": "VarianceScaling", "config": {"scale": 2.0, "mode": "fan_in", "distribution": "normal", "seed": null}}, "kernel_regularizer": null, "bias_regularizer": null, "activity_regularizer": null, "kernel_constraint": null, "bias_constraint": null}, "inbound_nodes": [[["dot_1", 0, 0, {}]]]}], "input_layers": [["input_1", 0, 0], ["input_2", 0, 0]], "output_layers": [["dense_2", 0, 0]]}, "keras_version": "2.3.1", "backend": "tensorflow"}
```
Comment (linusmichel) [inline similarity/src/resources/siamese-model-default.json:1]: Newline at EOF missing

**Review by linusmichel — CHANGES_REQUESTED** [review]
Functionality works as expected! 👍 
Added some minor formatting comments.

**Review by JonasPetry — COMMENTED** [review]
Functionality works for me, good job!

File: .github/workflows/dockerimage.yml [hunk 5]
```diff
@@ -58,3 +58,15 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v1
+    - uses: azure/docker-login@v1
+      with:
+        login-server: docker.pkg.github.com
+        username: ${GITHUB_ACTOR}
+        password: ${{ secrets.GITHUB_TOKEN }}
+    - name: Build and Push the athene-similarity Docker image
+      run: ./.github/workflows/scripts/dockerimage.sh "similarity"
```
[the remark this answers is not in the document]
Comment (clemenszuck) [reply .github/workflows/dockerimage.yml]: Fixed

File: similarity/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+COPY src/ src/
+
+# Run the image as a non-root user
+RUN groupadd -r similarity && useradd --no-log-init -r -g similarity similarity
+RUN mkdir -p /usr/src/app/logs/ && chown -R similarity:similarity /usr/src/app/
+VOLUME ["/usr/src/app/logs"]
+USER similarity
+
+EXPOSE 8000
+CMD gunicorn -b 0.0.0.0:8000 src.app
```
[the remark this answers is not in the document]
Comment (clemenszuck) [reply similarity/Dockerfile]: Fixed

File: similarity/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Listed dependencies as of 2020-06-24
+falcon==2.0.0
+gunicorn==20.0.4
+Keras==2.3.1
+numpy==1.16.5
+scikit-learn==0.21.3
+scipy==1.4.1
+tensorboard==2.1.0
+tensorflow==2.1.0
+
```
[the remark this answers is not in the document]
Comment (clemenszuck) [reply similarity/requirements.txt]: Fixed

File: similarity/requirements.txt [hunk 8]
```diff
@@ -0,0 +1,10 @@
+# Listed dependencies as of 2020-06-24
```
[the remark this answers is not in the document]
Comment (clemenszuck) [reply similarity/requirements.txt]: Fixed

File: similarity/src/NetworkTrainingResource.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+        doc = json.load(req.stream)
+        if "embeddingPairs" not in doc:
+            self.__logger.error("{} ({})".format(requireEmbeddingPairs.title, requireEmbeddingPairs.description))
+            raise requireEmbeddingPairs
+
+        embeddingPairs: List[EmbeddingsPair] = list(map(lambda dict: EmbeddingsPair.from_dict(dict), doc['embeddingPairs']))
+
+        #input dimension can be adapted here
+        self.__logger.info("Building model.")
+        self.__siameseNetwork.build_siamese_model((1024, 1))
```
[the remark this answers is not in the document]
Comment (clemenszuck) [reply similarity/src/NetworkTrainingResource.py]: Good point, it definitely could be useful! I changed it.

File: similarity/src/SimilarityResource.py (hunk 1 above)
Comment (clemenszuck) [reply similarity/src/SimilarityResource.py]: Fixed

File: similarity/src/entities.py (hunk 2 above)
Comment (clemenszuck) [reply similarity/src/entities.py]: Fixed

File: similarity/src/errors.py (hunk 3 above)
Comment (clemenszuck) [reply similarity/src/errors.py]: Fixed

File: similarity/src/resources/siamese-model-default.json (hunk 4 above)
Comment (clemenszuck) [reply similarity/src/resources/siamese-model-default.json]: Fixed

**Review by JonasPetry — APPROVED** [review]
Re-Approve

**Pull request closed without merging**